### PR TITLE
Convert AssignOutParametersAboveReturnTests to the new test framework

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs [hunk 1]
```diff
@@ -143,30 +140,31 @@ bool Try(out int i)
...
         public async Task TestMultiple()
         {
-            await TestInRegularAndScriptAsync(
+            await VerifyCS.VerifyCodeFixAsync(
 @"class C
 {
     char M(out int i, out string s)
     {
-        [|return 'a';|]
+        {|CS0177:{|CS0177:return 'a';|}|}
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: Interesting, duplicate diagnostics need nested annotation?

File: src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs [hunk 2]
```diff
@@ -42,6 +42,14 @@ protected override void TryRegisterFix(CodeFixContext context, Document document
         {
             var generator = editor.Generator;
 
+            if (exprOrStatement is LocalFunctionStatementSyntax { ExpressionBody: { } localFunctionExpressionBody })
```
Comment (mavasani) [inline src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs]: Tagging @CyrusNajmabadi for the product change, as I believe he added these code fixes.

Comment (sharwell) [reply src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs]: The results here were totally wrong. The code fix made a change in an unrelated method, with two impacts:

1. The original diagnostic was still reported in the same place, with a code fix offered
2. The iterative application of code fixes was an infinite loop

There was no way to port the test of local functions without this change.

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs [hunk 3]
```diff
@@ -3,29 +3,25 @@
...
 using Microsoft.CodeAnalysis.Test.Utilities;
 using Xunit;
+using VerifyCS = Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions.CSharpCodeFixVerifier<
+    Microsoft.CodeAnalysis.Testing.EmptyDiagnosticAnalyzer,
+    Microsoft.CodeAnalysis.CSharp.AssignOutParameters.AssignOutParametersAboveReturnCodeFixProvider>;
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.AssignOutParameters
 {
-    public partial class AssignOutParametersAboveReturnTests : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
+    public class AssignOutParametersAboveReturnTests
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: nice.

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs [hunk 4]
```diff
@@ -3,29 +3,25 @@
...
     {
-        internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderAndFixer(Workspace workspace)
-            => (null, new AssignOutParametersAboveReturnCodeFixProvider());
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: 👍

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs [hunk 5]
```diff
@@ -3,29 +3,25 @@
...
         public async Task TestForSimpleReturn()
         {
-            await TestInRegularAndScriptAsync(
+            await VerifyCS.VerifyCodeFixAsync(
 @"class C
 {
     char M(out int i)
     {
-        [|return 'a';|]
+        {|CS0177:return 'a';|}
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: awesome.  really like the declarative nature of this.

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs [hunk 6]
```diff
@@ -70,29 +66,30 @@ char M(out int i)
...
         public async Task TestMissingWhenVariableAssigned()
         {
-            await TestMissingInRegularAndScriptAsync(
-@"class C
+            var code = @"class C
 {
     char M(out int i)
     {
         i = 0;
-        [|return 'a';|]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: nice.  this never made much sense anyways (for a non-refactoring).

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: Interested in hearing what this is about as well.

File: src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs]: yikes!

File: src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs [hunk 7]
```diff
@@ -42,6 +42,14 @@ protected override void TryRegisterFix(CodeFixContext context, Document document
...
+            {
+                // Expression-bodied local functions report CS0177 on the method name instead of the expression.
+                // Reassign exprOrStatement so the code fix implementation works as it does for other expression-bodied
+                // members.
+                exprOrStatement = localFunctionExpressionBody.Expression;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersAboveReturnCodeFixProvider.cs]: 👍

**Comment by CyrusNajmabadi** [comment]
@sharwell because of the largely mechanical nature of the test changes, it's somewhat tedious to review.  If possible, would you be willing to go through and call out anything of particular interest (in the test changes) to make the review more expedient?    That way we can focus on the important bits, skim the rest, and get through a lot faster.

Thanks!

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: All diagnostics must be declared. There is no such thing as a "nested" annotation; this is just a visual artifact caused by one diagnostic's span being a superset of another.

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs [hunk 8]
```diff
@@ -348,14 +348,21 @@ void X()
...
-            await TestMissingInRegularAndScriptAsync(
+            await VerifyCS.VerifyCodeFixAsync(
 @"class C
 {
     void X()
     {
-        char D(out int i) => [|'a';|]
+        char {|CS0177:D|}(out int i) => 'a';
+        D(out _);
+    }
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: 📝 Prior to my change, the code fix for this produced this method:

```csharp
void X()
{
    i = 0;
    char D(out int i) => 'a';
    D(out _);
}
```

**Comment by sharwell** [comment]
> @sharwell because of the largely mechanical nature of the test changes, it's somewhat tedious to review. If possible, would you be willing to go through and call out anything of particular interest (in the test changes) to make the review more expedient? That way we can focus on the important bits, skim the rest, and get through a lot faster.

@CyrusNajmabadi I've now commented on the PRs sent to date. For future changes, I'll submit the PR as a draft, and include the annotations before marking it ready for review.

File: src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: >  one diagnostic's span being a superset of another.

but in this case, the spans are identical.  

1. does this indicate an issue with the analyzer (producing duplicate diagnostics)
2. if they're identical, does the test need to declare both?

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/AssignOutParameters/AssignOutParametersAboveReturnTests.cs]: There is no analyzer here. The compiler reported both diagnostics.

**Pull request merged**